subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -0,0 +1,346 @@
+    def "re-configuring #change in nested bean after the task started executing has no effect"() {
+        def fixture = new NestedBeanTestFixture()
+        fixture.prepareInputFiles()
+        buildFile << fixture.taskWithNestedProperty()
+        buildFile << """
+            taskWithNestedProperty.doLast {
+                bean = secondBean
+            }
+        """
+
+        when:
+        fixture.runTask()
+
+        then:
+        executedAndNotSkipped(fixture.task)
+
+        when:
+        fixture.changeFirstBean(change)
+        fixture.runTask()
+
+        then:
+        executedAndNotSkipped(fixture.task)
+
+        when:
+        fixture.changeSecondBean(change)
+        fixture.runTask()
+
+        then:
+        skipped(fixture.task)
+
+        where:
+        change << ['inputProperty', 'inputFile', 'outputFile']
+    }
+
+    class NestedBeanTestFixture {
+        def firstInputFile = 'firstInput.txt'
+        def firstOutputFile = 'build/firstOutput.txt'
+        def secondInputFile = 'secondInput.txt'
+        def secondOutputFile = 'build/secondOutput.txt'
+
+        def task = ':taskWithNestedProperty'
+
+        def inputProperties = [
+            first: 'first',
+            second: 'second'
+        ]
+        def inputFiles = [
+            first: file(firstInputFile),
+            second: file(secondInputFile)
+        ]
+        def outputFiles = [
+            first: file(firstOutputFile),
+            second: file(secondOutputFile)
+        ]
+
+        def changes = [
+            inputProperty: { String property ->
+                inputProperties[property] = inputProperties[property] + ' changed'
+            },
+            inputFile: { String property ->
+                inputFiles[property] << ' changed'
+            },
+            outputFile: { String property ->
+                outputFiles[property] << ' changed'
+            }
+        ]
+
+        def changeFirstBean(String change) {
+            changes[change]('first')
+        }
+
+        def changeSecondBean(String change) {
+            changes[change]('second')
+        }
+
+        def prepareInputFiles() {
+            file(firstInputFile).text = "first input file"
+            file(secondInputFile).text = "second input file"
+        }
+
+        def runTask() {
+            result = executer.withTasks(task, '-PfirstInput=' + inputProperties.first, '-PsecondInput=' + inputProperties.second).run()
+        }
+
+        String taskWithNestedProperty() {
+            """
+            class TaskWithNestedProperty extends DefaultTask {
+                @Nested
+                Object bean
+    
+                @OutputFile
+                RegularFileProperty outputFile = newOutputFile()
+    
+                @TaskAction
+                void writeInputToFile() {
+                    outputFile.getAsFile().get().text = bean
+                    bean.doStuff()
+                }
+            }
+    
+            class NestedBeanWithInput {
```

Comment:
This is not really `WithInput`, I think we should just call these guys `NestedBean` and `OtherNestedBean`.

Reply:
Renamed